src/tfr_mission_control/src/tfr_mission_control/mission_control.cpp

```diff
@@ -0,0 +1,409 @@
+     * seconds)
+     *
+     * I can't set the watchdog too long, or else there is an unacceptable delay
+     * in stopping the motors. So I make it a requirement that our laptop have a
+     * reasonable key delay set. 
+     *
+     * QT provides some built in functionality for this, but it's rubbish, and I
+     * couldn't get fine enough control to meet our response requirements.
+     * */
+    bool MissionControl::eventFilter(QObject* obj, QEvent* event)
+    {
+        if (event->type()==QEvent::KeyPress && teleopEnabled) {
+            QKeyEvent* key = static_cast<QKeyEvent*>(event);
+            int k = key->key();
+            //repeated key?
+            if(k == lastKey && motorKill->isActive())
+            {
+                //extend the watchdog
+                motorKill->start(MOTOR_INTERVAL);
+                return true;
+            }
+            //process the key and start watchdog
+            if (k == Qt::Key_W)
+            {
+                motorKill->start(MOTOR_INTERVAL);
+                lastKey =k;
+                performTeleop(tfr_utilities::TeleopCode::FORWARD);
+            }
+            else if(k == Qt::Key_S)
+            {
+                motorKill->start(MOTOR_INTERVAL);
+                lastKey =k;
+                performTeleop(tfr_utilities::TeleopCode::BACKWARD);
+            }
+            else if(k == Qt::Key_A)
+            {
+                motorKill->start(MOTOR_INTERVAL);
+                lastKey =k;
+                performTeleop(tfr_utilities::TeleopCode::LEFT);
+            }
+            else if(k == Qt::Key_D)
+            {
+                motorKill->start(MOTOR_INTERVAL);
+                lastKey =k;
+                performTeleop(tfr_utilities::TeleopCode::RIGHT);
+            }
+            //consume the key
+            return true;
+        }
+        else {
+            //unrecognized event type, pass it on
+            return QObject::eventFilter(obj, event);
+        }
+    }
+
+    /* ========================================================================== */
+    /* Callbacks                                                                  */
+    /* ========================================================================== */
+    /*
+     * Callback for our status subscriber this happens in the "ros" thread so I
+     * need to communicate to the GUI in a safe signal/slot combination.
+     *
+     * This triggers our custom emitStatus signal, which triggers the built in
+     * text update in the text box, a seperate timer event is constantly
+     * scrolling the window on a set fast interval.
+     * */
+    void MissionControl::updateStatus(const tfr_msgs::SystemStatusConstPtr &status)
+    {
+        auto str = getStatusMessage(static_cast<StatusCode>(status->status_code), status->data);
+        //need to wrap in signal safe "Q" object
+        QString msg = QString::fromStdString(str);
+        emit emitStatus(msg);
```

Comment:
Just 'cause I'm curious, is this some sort of special Qt preprocessor thing?

Reply:
It's sends the string to the next event in the chain of events to process

Reply:
Yeah, I was just wondering about the `emit emitStatus(msg);`, that seems like preprocessor stuff unless there's some weird `emit` keyword in C++ I don't know about, hahaha.